File: backend/node.js-typescript/src/infrastructure/core/query-handler/collection.query.result.ts
/**
 * Represents a query result containing a collection of items of type T.
 * @typeparam T - The type of items contained in the collection.
 */
export class CollectionQueryResult<T> {
    /**
     * The collection of items.
     */
    items: T[];

    /**
     * The total number of items in the collection.
     */
    totalItems: number;

    /**
     * Initializes a new instance of the CollectionQueryResult class.
     * @param items - The collection of items.
     * @param totalItems - The total number of items in the collection.
     */
    constructor(items: T[], totalItems?: number) {
        if (items == null) {
            throw new Error("Items collection cannot be null.");
        }

        this.items = items;
        this.totalItems = totalItems !== undefined ? totalItems : items.length;
    }
}

File: backend/dotnet/Framework/Framework.Core/Queries/CollectionQueryResult.cs
namespace Framework.Core.Queries;

/// <summary>
///     Represents a query result containing a collection of items of type T.
/// </summary>
/// <typeparam name="T">The type of items contained in the collection.</typeparam>
public class CollectionQueryResult<T> : IQueryResult
{
    /// <summary>
    ///     Initializes a new instance of the <see cref="CollectionQueryResult{T}" /> class.
    /// </summary>
    /// <param name="items">The collection of items.</param>
    /// <param name="totalItems">The total number of items in the collection.</param>
    public CollectionQueryResult(IEnumerable<T> items, int totalItems)
    {
        Items = items ?? throw new ArgumentNullException(nameof(items), "Items collection cannot be null.");
        TotalItems = totalItems;
    }

    /// <summary>
    ///     Initializes a new instance of the <see cref="CollectionQueryResult{T}" /> class.
    /// </summary>
    /// <param name="items">The collection of items.</param>
    public CollectionQueryResult(IEnumerable<T> items) : this(items, items.Count())
    {
    }

    /// <summary>
    ///     Gets or sets the total number of items in the collection.
    /// </summary>
    public int TotalItems { get; set; }

    /// <summary>
    ///     Gets or sets the collection of items.
    /// </summary>
    public IEnumerable<T> Items { get; set; }
}